File: [C]hashmap/[C]hashmap/13-roman-to-integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        char_val = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
        res = 0
        for i, c in enumerate(s):
            if i + 1 < len(s) and char_val[c] < char_val[s[i + 1]]:
                res -= char_val[c]
            else:
                res += char_val[c]
        return res
        
# time O(n), due to traverse, n is the length of input string
# space O(1), hashmap's size is O(7)
# using hashmap and store val

File: [C]hashmap/[C]hashmap/13-roman-to-integer.java
class Solution {
    public int romanToInt(String s) {
        HashMap<Character, Integer> charToWeight = new HashMap<>();
        charToWeight.put('I', 1);
        charToWeight.put('V', 5);
        charToWeight.put('X', 10);
        charToWeight.put('L', 50);
        charToWeight.put('C', 100);
        charToWeight.put('D', 500);
        charToWeight.put('M', 1000);
        int res = 0;
        for (int i = 0; i < s.length(); i++) {
            Character c = s.charAt(i);
            int w = charToWeight.get(c);
            if (i + 1 < s.length() && charToWeight.get(s.charAt(i + 1)) > w) {
                res -= w;
            } else {
                res += w;
            }
        }
        return res;
    }
}

// time O(n), due to traverse, n is the length of input string
// space O(1), hashmap's size is O(7)
// using hashmap and store val